### PR TITLE
Refactor remote store flow to support any path type

### LINKED ISSUE #12790: [Remote Store] Update remote Store flows to support any path type with backward compatibility
### Is your feature request related to a problem? Please describe

With optimised prefix pattern for remote store path (as mentioned in https://github.com/opensearch-project/OpenSearch/issues/12567), we need to ensure that we are able to allow new indexes to use the optimised remote store path type and at the same time be able to resolve the existing prefix pattern (Fixed remote store path type). 

### Describe the solution you'd like

As part of this task, we intend to make the necessary code changes to allow new indexes to use any remote store path type based on the cluster setting `cluster.remote_store.index.path.prefix.type`. We have added a remote store custom data in index metadata that is used to associate the path type for an index. When the custom data is absent, we would treat the path type to be Fixed.  

### Related component

Storage:Remote

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ashking94**
### Description
This PR is a sub task of feature request https://github.com/opensearch-project/OpenSearch/issues/12567 and being tracked in meta issue - https://github.com/opensearch-project/OpenSearch/issues/12589.

In the prior PR https://github.com/opensearch-project/OpenSearch/pull/12607 & https://github.com/opensearch-project/OpenSearch/pull/12753, we have introduced remote store path type in the customData field in IndexMetadata and handled association of path types of indexes during snapshot restore. In this PR, we are refactoring existing remote store flows to add support for using any remote store path type with backward compatibility. 

In brief, we are doing follow things -
1. We are replacing all places where we were deriving paths manually (like RemoteSegmentStoreDirectoryFactory.java, RemoteStoreLockManagerFactory.java, TranslogTransferManager & BlobStoreRepository.java - snapshots related) with deriving path using RemoteStorePathType.
2. We provide implementation for the existing fixed prefix patten, and handle the wiring also.
3. Write UTs for some new classes. 

### Related Issues
Resolves #12790

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- ~[ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))~
- ~[ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ashking94** [comment]
> testConcurrentDecommissionAction



> ❕ Gradle check result for [21636af](https://github.com/opensearch-project/OpenSearch/commit/21636af8b01ff7eb1a63fb5370d4174f15b38e1b): [UNSTABLE](https://build.ci.opensearch.org/job/gradle-check/35575/)
> 
> * **TEST FAILURES:**
> 
> ```
>       2 org.opensearch.cluster.coordination.AwarenessAttributeDecommissionIT.testConcurrentDecommissionAction
> ```
> 
> Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

Flaky test - https://github.com/opensearch-project/OpenSearch/issues/12197

File: server/src/test/java/org/opensearch/index/store/RemoteSegmentStoreDirectoryTests.java [hunk 1]
```diff
@@ -697,13 +698,20 @@ public void testCleanupAsync() throws Exception {
...
         );
-        when(remoteSegmentStoreDirectoryFactory.newDirectory(any(), any(), any())).thenReturn(remoteSegmentDirectory);
+        when(remoteSegmentStoreDirectoryFactory.newDirectory(any(), any(), any(), any())).thenReturn(remoteSegmentDirectory);
         String repositoryName = "test-repository";
         String indexUUID = "test-idx-uuid";
         ShardId shardId = new ShardId(Index.UNKNOWN_INDEX_NAME, indexUUID, Integer.parseInt("0"));
-
-        RemoteSegmentStoreDirectory.remoteDirectoryCleanup(remoteSegmentStoreDirectoryFactory, repositoryName, indexUUID, shardId);
-        verify(remoteSegmentStoreDirectoryFactory).newDirectory(repositoryName, indexUUID, shardId);
+        RemoteStorePathType pathType = RemoteStorePathType.FIXED;
```
Comment (BhumikaSaini-Amazon) [inline server/src/test/java/org/opensearch/index/store/RemoteSegmentStoreDirectoryTests.java:705]: Do we need randomization for the pathType value?

File: server/src/test/java/org/opensearch/index/store/lockmanager/RemoteStoreLockManagerFactoryTests.java [hunk 2]
```diff
@@ -48,6 +49,7 @@ public void testNewLockManager() throws IOException {
         String testRepository = "testRepository";
         String testIndexUUID = "testIndexUUID";
         String testShardId = "testShardId";
+        RemoteStorePathType pathType = RemoteStorePathType.FIXED;
```
Comment (BhumikaSaini-Amazon) [inline server/src/test/java/org/opensearch/index/store/lockmanager/RemoteStoreLockManagerFactoryTests.java:52]: Do we need randomization for the pathType value?

**Review by BhumikaSaini-Amazon — CHANGES_REQUESTED** [review]

File: server/src/test/java/org/opensearch/index/store/RemoteSegmentStoreDirectoryTests.java (hunk 1 above)
Comment (ashking94) [reply server/src/test/java/org/opensearch/index/store/RemoteSegmentStoreDirectoryTests.java:705]: I have added this.

**Comment by ashking94** [comment]
> ❌ Gradle check result for [e21ec8f](https://github.com/opensearch-project/OpenSearch/commit/e21ec8fd3f37791557ecbf2c73af3a22015fa87e): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/35592/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test - https://github.com/opensearch-project/OpenSearch/issues/10558 and https://github.com/opensearch-project/OpenSearch/issues/12197

**Review by BhumikaSaini-Amazon — APPROVED** [review]

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java [hunk 3]
```diff
@@ -75,6 +78,8 @@ public class RemoteFsTranslog extends Translog {
     private static final int REMOTE_DELETION_PERMITS = 2;
     private static final int DOWNLOAD_RETRIES = 2;
     public static final String TRANSLOG = "translog";
+    public final static String METADATA_DIR = "metadata";
+    public final static String DATA_DIR = "data";
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:82]: can we unify tlog and segments `data` and `metadata` prefix names ?

Comment (ashking94) [reply server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:82]: Let me do this, I was thinking of creating an enum for the same.

Comment (ashking94) [reply server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java]: Addressed your comment.

**Comment by ashking94** [comment]
> ❕ Gradle check result for [d9e12db](https://github.com/opensearch-project/OpenSearch/commit/d9e12db0af8f6a0798d02ad46e9460d18d774e7d): [UNSTABLE](https://build.ci.opensearch.org/job/gradle-check/35758/)
> 
> * **TEST FAILURES:**
> 
> ```
>       2 org.opensearch.index.shard.RemoteStoreRefreshListenerTests.classMethod
>       1 org.opensearch.smoketest.SmokeTestMultiNodeClientYamlTestSuiteIT.test {yaml=search.aggregation/20_terms/string profiler via global ordinals}
>       1 org.opensearch.index.shard.RemoteStoreRefreshListenerTests.testRefreshSuccessAfterFailureInFirstAttemptAfterSnapshotAndMetadataUpload
> ```
> 
> Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

flaky tests - https://github.com/opensearch-project/OpenSearch/issues/8131 and https://github.com/opensearch-project/OpenSearch/issues/9332

**Review by gbbafna — APPROVED** [review]

**Comment by ashking94** [comment]
Auto backport failed, raising manual backport.

**Pull request merged**